### PR TITLE
Cleanup some FIXMEs in ainvs

### PR DESCRIPTION
**Pull request by MitchellBuckley**
This is an attempt to remove some `FIXME`s from ainvs.

* move lots of lemmas up to `lib`.
* replace `cte_wp_at_eqD` with `cte_wp_at_norm` (equal lemmas)
* `pd_shifting_gen` generalises `pd_shifting'` in 2 architectures
* remove some redundant crunch lemmas

I could have kept going with this to make a much larger pull-request, but thought I should check that I'm moving things to the right places.

Testboard here: https://bamboo.ts.data61.csiro.au/browse/L4V-GHTESTBOARD60-1

### CODE REVIEW

File: lib/LemmaBucket.thy
Comment (Xaphiosis) [inline lib/LemmaBucket.thy:514]: This one is a bit specific for lib, but there isn't an ARM*-common area. It might be worth considering which of these could go into Word_Lib, but maybe as a future next step.

**Review by Xaphiosis — APPROVED** [review]
Nice work.
I'm a bit torn with the MCS situation on how much cleanup master can take before the merges become an issue, but I think this kind of cleanup is relatively safe.

File: lib/LemmaBucket.thy
Comment (MitchellBuckley) [reply lib/LemmaBucket.thy:514]: If it's too specific for lib, there's probably somewhere in ainvs that I could find a place for it. I'll try and do that.

And yes, not sure which of these can move up to Word_Lib or similar, but I'm glad you're happy to leave that for a future step.

**Comment by MitchellBuckley** [comment]
> Nice work.
> I'm a bit torn with the MCS situation on how much cleanup master can take before the merges become an issue, but I think this kind of cleanup is relatively safe.

Yeah, good point. Do you have any concerns about impact on MCS @mktnk3 @lsf37 ?

**Comment by mktnk3** [comment]
Hmm. I’d think this kind of clean up makes sense if it contributes to reduce the gap between master and MCS. Doing this only to master is likely to create more work in the future. This is not to stop doing this, but we really need to apply the same to MCS as part of this effort.

The corres rename task came up from MCS because we were naming new corres rules (MCS specific ones) consistently, while the existing ones were not, and that was causing some confusion to some. Victor started it on master for the purpose of staying in sync with master. So, the task is not really complete till it is applied to MCS.

**Review by mbrcknl — APPROVED** [review]
The clean-up looks good!

Not sure how to answer the question about merge issues, though. @lsf37 is probably the person who knows most about how well the branches are currently synced, and how to best keep them that way. But he seems quite overloaded at the moment.

**Comment by mbrcknl** [comment]
And yes, I agree with @mktnk3 that this will work best if we make an effort now to sync these changes to both branches.

**Comment by MitchellBuckley** [comment]
Ok, happy to apply parallel changes to MCS. I guess the obvious way would be to cherry-pick the commit and manually check that the changes are in sync. But I'll wait for Gerwin's input before I do that, and open to other ideas.

**Review by lsf37 — APPROVED** [review]
This is great! I’m not too worried about the merge impact with MCS on this one. Happy for it to go ahead.

**Comment by lsf37** [comment]
> Ok, happy to apply parallel changes to MCS. I guess the obvious way would be to cherry-pick the commit and manually check that the changes are in sync. But I'll wait for Gerwin's input before I do that, and open to other ideas.

I’d actually wait with that, unless the patch just applies as is (worth trying, but chances are low). Cherry-picking is good when it applies cleanly, but if you need to resolve conflicts during that, you’ll get another set of conflicts from those in the merge, i.e. we’d be doing the work twice. 

I should probably start looking into the next merge of master into rt, just currently a bit busy with setting up GitHub CI, so it’ll be a few weeks until I get there.

**Comment by mktnk3** [comment]
> I should probably start looking into the next merge of master into rt, just currently a bit busy with setting up GitHub CI, so it’ll be a few weeks until I get there.

It’s been about a year since the last merge, and master have changed to Isabelle2020 in the meantime. Do you think you can merge across Isabelle versions? If that’s possible and you think you can manage the changes here fine, I suppose I should leave it with you. Do you think the same for the corres rename as well?

**Comment by lsf37** [comment]
The last master/rt merge was also across Isabelle versions. It's messy, but it's doable. The corres rename is probably also better to defer until after that, yes. At least if I get to it in the next 2 months :-)

**Comment by mktnk3** [comment]
> The last master/rt merge was also across Isabelle versions. It's messy, but it's doable. The corres rename is probably also better to defer until after that, yes. At least if I get to it in the next 2 months :-)

Ah I see. I probably didn’t read the log close enough. I am not too sure about the two months timeframe :), but I suppose I can leave it with you then.

As for the corres renames, I am concerned about those that are already applied to master. Merging should take care of reflecting those in MCS (manually?), I would think? But wouldn’t it easier to run the same script on MCS before merging? But, either way, if merging is something that can happen soonish, we could leave it to revisit at that point maybe.

As for the new MCS ones, we have been writing them “correctly” for a while now, although there may be some older ones left to be updated.

**Comment by lsf37** [comment]
> As for the corres renames, I am concerned about those that are already applied to master. Merging should take care of reflecting those in MCS (manually?), I would think? But wouldn’t it easier to run the same script on MCS before merging? But, either way, if merging is something that can happen soonish, we could leave it to revisit at that point maybe.

It's similar to this PR: if the same rename patch applies cleanly without conflict resolution, applying to to `rt` would be fine and would not create additional work, but if we recreate it manually and slightly differently (or we apply it and have to do conflict resolution), then we generate guaranteed conflicts in the merge. This would mean we're doing application/conflict resolution twice.
 
> As for the new MCS ones, we have been writing them “correctly” for a while now, although there may be some older ones left to be updated.

That's excellent, this should reduce the difference by a good bit.

**Pull request merged**
